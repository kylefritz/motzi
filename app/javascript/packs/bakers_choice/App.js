import React, { useState } from 'react'
import axios from 'axios'

import User from './User'

export default function BakersChoice() {
  const { haventOrdered, menu } = gon
  const [users, setUsers] = useState(haventOrdered)

  const handlePickBread = ({ userId, itemId }) => {
    const order = { userId, itemId }
    console.log("create order", order, 'for menu:', menu.id)
    axios.post(`/admin/menus/bakers_choice.json`, order).then(({ data: users }) => {
      setUsers(users)
    })
  }

  return (
    <table>
      <thead>
        <tr>
          <th>First Name</th>
          <th>Last Name</th>
          <th>Credits Remaining</th>
          <th>Breads per Week</th>
          <th>Tuesday Pickup?</th>
          <th>Asign Bread</th>
        </tr>
      </thead>
      <tbody>
        {users.map(u => <User key={u.id} {...u} menu={menu} onPickBread={handlePickBread} />)}
      </tbody>
    </table>
  )
}
