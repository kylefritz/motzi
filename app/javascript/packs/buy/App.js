import React, { useEffect, useState } from "react";
import axios from 'axios'
import * as Sentry from '@sentry/browser'
import _ from 'lodash'
import queryString from "query-string";

import Choice from './Choice'
import Payment from './Payment'

export default function Buy({onComplete, user: passedUser=null}) {
  const [credits, setCredits] = useState();
  const [price, setPrice] = useState();
  const [breadsPerWeek, setBreadsPerWeek] = useState();
  const [user, setUser] = useState(passedUser);
  const [error, setError] = useState();
  const [receipt, setReceipt] = useState();

  // what is the current user?
  useEffect(() => {
    if(user) {
      return;
    }
    const { uid } = queryString.parse(location.search);
    const params = { uid };
    axios.get("/menu.json", { params }).then(({ data: {user} }) => {
      if(user)
      {
        setUser(user);
        Sentry.configureScope(scope => scope.setUser(user));
      }else{
        setError("We can't load your user account");
      }
    })
    .catch(error => {
      console.error("cant load user from menu page", error);
      Sentry.captureException(error);
      setError("We can't load your user account");
    });
  }, []);

  const handleChoose = ({ credits, price, breadsPerWeek }) => {
    setCredits(credits);
    setPrice(price);
    setBreadsPerWeek(breadsPerWeek)
    console.log("selelcted", credits, "for", "price");
  };

  const handlePriceChanged = (priceString) =>{
    if(priceString){
      const newPrice = parseFloat(priceString);

      if (_.isFinite(newPrice) && newPrice > 0) {
        setPrice(newPrice);
      }
    }
  }

  const handleCardToken = ({ token }) => {
    console.log("card token=", token, { credits, price });

    // send stripe token to rails to complete purchase
    axios.post("/credit_items.json", { uid: user.id, token: token.id, price, credits, breadsPerWeek })
      .then(({ data }) => {
        const {creditItem} = data
        console.log("bought credits", data);

        setReceipt(creditItem.stripeReceiptUrl);

        if(onComplete)
        {
          onComplete(creditItem);
        }
      })
      .catch(error => {
        console.error("cant buy credits order", error);
        window.alert("There was a problem buying credits.");
        Sentry.captureException(error);
      });
  };

  const handlePaymentResult = ({ token }) => {
    // with payment request, the price sent to stripe
    console.log("paymentResult token=", token);

    // TODO: send completed payment request to rails
  };

  if(receipt)
  {
    return (
      <div className="alert alert-success" role="alert">
        <h2>Thanks for supporting Motzi!!</h2>
        <p className="text-center my-3">
          <a href={receipt} target="blank">
            Here's your receipt
          </a>
        </p>
      </div>
    );
  }

  if(error)
  {
    return <div className="text-center my-3">{error}</div>;
  }

  if (!user) {
    return (
      <h6 className="text-center my-3">
        Loading user for subscription renewal...
      </h6>
    );
  }

  return (
    <div className="alert alert-info padding-x-mobile-5px" role="alert">
      <h2 className="mb-3" style={{fontSize: '1.8rem'}}>Buy credits</h2>

      <h6>6-month</h6>
      <div>
        <Choice breadsPerWeek={1.0} credits={26} price={6.5} total={169} onChoose={handleChoose}/>
        <Choice breadsPerWeek={0.5} credits={13} price={7.0} total={91} onChoose={handleChoose}/>
      </div>

      <h6>3-month</h6>
      <div>
        <Choice breadsPerWeek={1.0} credits={13} price={7.0} total={91} onChoose={handleChoose}/>
        <Choice breadsPerWeek={0.5} credits={6}  price={7.5} total={46} onChoose={handleChoose} />
      </div>
      {
        (credits && price) && (
          <>
            <h4>Payment amount</h4>
            <p>
              <small>
                These are suggested prices based on the our costs. We want everyone to
                have access to healthy food in this time of crisis and beyond,
                regardless of ability to pay. If you are out of work or otherwise in a
                challenging financial situation please pay what you can or nothing at
                all. If you have the means to pay more we would welcome your support
                in providing for the community and ensuring our continued stability as
                a small business.
              </small>
            </p>
            <div className="input-group mb-3">
              <div className="input-group-prepend">
                <span className="input-group-text" id="you-pay">You pay $</span>
              </div>
              <input type="number" min="1" max="250" className="form-control" placeholder="Price" aria-label="Price"
                    aria-describedby="you-pay" defaultValue={price} onBlur={(e) => handlePriceChanged(e.target.value)} />
            </div>
            <Payment
              credits={credits}
              price={price}
              stripeApiKey={gon.stripeApiKey}
              onCardToken={handleCardToken}
              onPaymentResult={handlePaymentResult}
            />
          </>
        )
      }
    </div>
  );
}
