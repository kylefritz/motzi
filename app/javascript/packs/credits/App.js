import React from 'react'
import axios from 'axios'
import * as Sentry from '@sentry/browser'
import _ from 'lodash'

export default function App() {
  const memoRef = React.createRef()
  const quantityRef = React.createRef()
  const weeksRef = React.createRef()
  const userId = _.get(location.pathname.match(/users\/(.*)/), 1)

  const handleSubmit = (event) => {
    event.preventDefault()
    const credit = {
      memo: memoRef.current.value,
      quantity: quantityRef.current.value,
      goodForWeeks: weeksRef.current.value,
      userId
    }
    console.info(credit)
    axios.post(`/admin/credit_items.json`, credit).then(({ data }) => {
      // reload page
      document.location = document.location
    }).catch((error) => {
      console.error("couldn't create credit", error)
      Sentry.captureException(error)
      alert("Couldn't create credit?!")
    })
  }

  return (
    <details>
      <summary>Add credit</summary>
      <form onSubmit={handleSubmit.bind(this)}>
        <fieldset>
          <ol>
            <li className="string input optional stringish">
              <label htmlFor="credit_entry_memo" className="label">Memo</label>
              <input type="text" ref={memoRef} />
            </li>

            <li className="number input optional numeric stringish">
              <label htmlFor="credit_entry_quantity" className="label">Quantity</label>
              <input step="any" type="number" ref={quantityRef} />
            </li>

            <li className="number input optional numeric stringish">
              <label htmlFor="credit_entry_good_for_weeks" className="label">Good for weeks</label>
              <input step="any" type="number" ref={weeksRef} />
            </li>
          </ol>
        </fieldset>
        <fieldset>
          <ol>
            <li className="action input_action " id="credit_entry_submit_action">
              <input type="submit" value="Add credit" />
            </li>
          </ol>
        </fieldset>
      </form >
    </details>
  )
}
