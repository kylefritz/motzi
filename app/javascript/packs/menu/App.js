import React from 'react'
import axios from 'axios'
import * as Sentry from '@sentry/browser';
import queryString from 'query-string'

import Order from './Order.js'
import Menu from './Menu.js'

export default class App extends React.Component {
  componentDidMount() {
    const { uid } = queryString.parse(location.search)
    axios.get('/menu.json', { params: { uid } }).then(({ data }) => {
      this.setState(data) // expect: menu, user, order
      const { user } = data
      Sentry.configureScope((scope) => scope.setUser(user))
    }).catch((error) => {
      console.error("cant load menu", error)
      Sentry.captureException(error)
      this.setState({ error: "We can't load the menu" })
    })
  }

  handleCreateOrder(order) {
    if (this.state.order) {
      window.alert("Weird. This web site thinks you already placed an order. Refresh the page?!")
      return
    }

    console.debug('creating order', order)
    axios.post('/orders.json', order).then(({ data }) => {
      this.setState(data) // expect: menu, user, order
      console.debug('created order', data)
    }).catch((error) => {
      console.error("cant create order", error);
      window.alert("There was a problem submitting your order.")
      Sentry.captureException(error)
    });
  }

  render() {
    const { menu, user, order, error } = this.state || {};
    if (error) {
      return <>
        <h2 className="mt-5">{error}</h2>
        <p className="text-center">Sorry. Maybe try again or try back later?</p>
      </>
    }

    if (!menu) {
      return <h2 className="mt-5">loading...</h2>
    }

    if (order) {
      return <Order {...{ user, order, menu }} />
    }

    return <Menu {...{ user, order, menu, onCreateOrder: this.handleCreateOrder.bind(this) }} />
  }
}
