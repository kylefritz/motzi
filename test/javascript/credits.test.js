import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

import App from 'credits/App'

test('snapshot', () => {
  const wrapper = shallow(<App />);
  expect(wrapper).toMatchSnapshot();
});
