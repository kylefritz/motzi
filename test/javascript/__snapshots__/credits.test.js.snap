// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<details>
  <summary>
    Add credit
  </summary>
  <form
    onSubmit={[Function]}
  >
    <fieldset>
      <ol>
        <li
          className="string input optional stringish"
        >
          <label
            className="label"
            htmlFor="credit_entry_memo"
          >
            Memo
          </label>
          <input
            type="text"
          />
        </li>
        <li
          className="number input optional numeric stringish"
        >
          <label
            className="label"
            htmlFor="credit_entry_quantity"
          >
            Quantity
          </label>
          <input
            step="any"
            type="number"
          />
        </li>
        <li
          className="number input optional numeric stringish"
        >
          <label
            className="label"
            htmlFor="credit_entry_good_for_weeks"
          >
            Good for weeks
          </label>
          <input
            step="any"
            type="number"
          />
        </li>
      </ol>
    </fieldset>
    <fieldset>
      <ol>
        <li
          className="action input_action "
          id="credit_entry_submit_action"
        >
          <input
            type="submit"
            value="Add credit"
          />
        </li>
      </ol>
    </fieldset>
  </form>
</details>
`;
